File: vito/core/src/main/java/com/facebook/fresco/vito/core/VitoImageRequest.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.core;

import android.content.res.Resources;
import com.facebook.cache.common.CacheKey;
import com.facebook.common.internal.Objects;
import com.facebook.fresco.vito.options.ImageOptions;
import com.facebook.fresco.vito.source.ImageSource;
import com.facebook.imagepipeline.request.ImageRequest;
import javax.annotation.Nullable;

public class VitoImageRequest {

  public final Resources resources;
  public final ImageSource imageSource;
  public final ImageOptions imageOptions;
  public final @Nullable ImageRequest finalImageRequest;
  public final @Nullable CacheKey finalImageCacheKey;

  public VitoImageRequest(
      Resources resources,
      ImageSource imageSource,
      ImageOptions imageOptions,
      @Nullable ImageRequest finalImageRequest,
      @Nullable CacheKey finalImageCacheKey) {
    this.resources = resources;
    this.imageSource = imageSource;
    this.imageOptions = imageOptions;
    this.finalImageRequest = finalImageRequest;
    this.finalImageCacheKey = finalImageCacheKey;
  }

  @Override
  public boolean equals(@Nullable Object obj) {
    if (this == obj) {
      return true;
    }
    if (obj == null || getClass() != obj.getClass()) return false;
    VitoImageRequest other = (VitoImageRequest) obj;
    return resources == other.resources
        && Objects.equal(imageSource, other.imageSource)
        && Objects.equal(imageOptions, other.imageOptions);
  }

  @Override
  public int hashCode() {
    int result = resources.hashCode();
    result = 31 * result + imageSource.hashCode();
    result = 31 * result + imageOptions.hashCode();
    return result;
  }
}

File: vito/core/src/main/java/com/facebook/fresco/vito/core/VitoImageRequest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.core

import android.content.res.Resources
import com.facebook.cache.common.CacheKey
import com.facebook.common.internal.Objects
import com.facebook.fresco.middleware.Extras
import com.facebook.fresco.middleware.HasExtraData
import com.facebook.fresco.vito.options.ImageOptions
import com.facebook.fresco.vito.source.ImageSource
import com.facebook.imagepipeline.request.ImageRequest
import kotlin.jvm.JvmField

class VitoImageRequest(
    @JvmField val resources: Resources,
    @JvmField val imageSource: ImageSource,
    @JvmField val imageOptions: ImageOptions,
    @JvmField val logWithHighSamplingRate: Boolean = false,
    @JvmField val finalImageRequest: ImageRequest?,
    @JvmField val finalImageCacheKey: CacheKey?,
    @JvmField val extras: MutableMap<String, Any> = mutableMapOf(),
) : HasExtraData {

  /** Simplify this once BaseCloseableImage and BaseProducerContext are converted to Kotlin */
  override fun <E> putExtra(key: String, value: E?) {
    if (value == null) {
      extras.remove(key)
    } else {
      extras[key] = value
    }
  }

  @Suppress("UNCHECKED_CAST") override fun <E> getExtra(key: String): E? = extras[key] as? E

  override fun <E> getExtra(key: String, valueIfNotFound: E?): E? = getExtra(key) ?: valueIfNotFound

  override fun getExtras(): Extras = extras

  override fun putExtras(extras: Extras) = extras.forEach { putExtra(it.key, it.value) }

  override fun equals(obj: Any?): Boolean {
    if (this === obj) {
      return true
    }
    if (obj == null || javaClass != obj.javaClass) return false
    val other = obj as VitoImageRequest
    return resources === other.resources &&
        Objects.equal(imageSource, other.imageSource) &&
        Objects.equal(imageOptions, other.imageOptions)
  }

  fun equalsIfHasImage(other: VitoImageRequest?, hasImage: Boolean): Boolean {
    if (this === other) {
      return true
    }
    if (other == null) {
      return false
    }
    return resources === other.resources &&
        Objects.equal(imageSource, other.imageSource) &&
        if (hasImage) imageOptions.equalsForActualImage(other.imageOptions)
        else Objects.equal(imageOptions, other.imageOptions)
  }

  override fun hashCode(): Int {
    var result = resources.hashCode()
    result = 31 * result + imageSource.hashCode()
    result = 31 * result + imageOptions.hashCode()
    return result
  }
}
